src/meta/src/barrier/backfill_order_control.rs

```diff
@@ -116,24 +119,40 @@ impl BackfillOrderState {
 // state transitions
 impl BackfillOrderState {
     pub fn finish_actor(&mut self, actor_id: ActorId) -> Vec<FragmentId> {
-        // Find the fragment_id of the actor.
-        if let Some(fragment_id) = self.actor_to_fragment_id.get(&actor_id)
-            // Decrease the remaining_actor_count of the operator.
-            // If the remaining_actor_count is 0, add the operator to the current_backfill_nodes.
-            && let Some(node) = self.current_backfill_nodes.get_mut(fragment_id)
-        {
-            assert!(node.remaining_actors.remove(&actor_id), "missing actor");
-            tracing::debug!(
-                actor_id,
-                remaining_actors = node.remaining_actors.len(),
-                fragment_id,
-                "finish_backfilling_actor"
-            );
-            if node.remaining_actors.is_empty() {
-                return self.finish_fragment(*fragment_id);
+        let Some(fragment_id) = self.actor_to_fragment_id.get(&actor_id) else {
+            tracing::error!(actor_id, "fragment not found for actor");
+            return vec![];
+        };
+        // Some nodes might be finished out of order, for instance
+        // if there's no data to backfill at all.
+        // in such cases, we should directly update it in remaining backfill nodes instead.
+        let (node, is_in_order) = match self.current_backfill_nodes.get_mut(fragment_id) {
+            Some(node) => (node, true),
+            None => {
+                let Some(node) = self.remaining_backfill_nodes.get_mut(fragment_id) else {
+                    tracing::error!(
+                        fragment_id,
+                        actor_id,
+                        "fragment not found in current_backfill_nodes or remaining_backfill_nodes"
+                    );
+                    return vec![];
+                };
+                (node, false)
             }
+        };
+
+        assert!(node.remaining_actors.remove(&actor_id), "missing actor");
+        tracing::debug!(
+            actor_id,
+            remaining_actors = node.remaining_actors.len(),
+            fragment_id,
+            "finish_backfilling_actor"
+        );
+        if node.remaining_actors.is_empty() && is_in_order {
+            self.finish_fragment(*fragment_id)
+        } else {
+            vec![]
```

Comment:
I added this section here for source backfill, since it may finish out of order (e.g. if there's nothing to backfill).

Reply:
tbh I don't fully understand the "out of order". Can you give an example?

Reply:
Will add more comments

Reply:
fixed https://github.com/risingwavelabs/risingwave/commit/909444cb6254140f3e3d2ef8464517a2f0912a5b